File: ASSAULT_COURSE/LONG_METHOD/src/com/codemanship/refactor/longmethod/Account.java
package com.codemanship.refactor.longmethod;

import java.util.Calendar;
import java.util.List;
import java.util.ArrayList;

public class Account {
	
	private float balance = 0;
	
	private List transactions = new ArrayList();

	private String lastDebitDate;
	
	private float owing = 0;

	public float getBalance() {
		return balance;
	}

	public void credit(float amount) {
		balance += amount;
	}

	public void debit(float amount) {
		
		// deduct amount from balance
		balance -= amount;
		
		// record transaction
		transactions.add(new Transaction(true, amount));
		
		// update last debit date
		Calendar calendar = Calendar.getInstance();
		
		lastDebitDate = calendar.get(Calendar.DATE) + "/" +
		calendar.get(Calendar.MONTH) + "/" + 
		calendar.get(Calendar.YEAR);
	}

	public Transaction getLastTransaction() {
		return (Transaction)transactions.get(transactions.size()  -1);
	}

	public String getLastDebitDate() {
		return lastDebitDate;
	}
}

File: ruby/duplicate_code/account.rb
require "date"

class Account
  attr_reader :balance, :transactions, :last_transaction_date

  def initialize()
    @balance = 0
    @transactions = []
  end

  def credit(amount)
    record_transaction(amount, false)
  end

  def debit(amount)
    record_transaction(-amount, true)
  end

  def last_transaction
    @transactions.last
  end

  private
    def record_transaction(amount, is_debit)
      @balance += amount
      @transactions << Transaction.new(is_debit, amount)
      @last_transaction_date = Date.today
    end

end
